### PR TITLE
fix: boolean slicing with non-packed arrays

### PR DESCRIPTION
**Pull request by agoose77**
In fixing #2214, I missed the fact that `to_ListOffsetArray64(True)` provides no guarantee about the _length_ of the content. This PR fixes that oversight, by using `ak._do.flatten` which ensures the length of the result is correct.

### CODE REVIEW

File: src/awkward/_slicing.py [hunk 1]
```diff
@@ -522,7 +527,9 @@ def _normalise_item_bool_to_int(item: Content, backend: Backend) -> Content:
     ):
         if item_backend.nplike.known_data:
             if isinstance(item_backend.nplike, Jax):
-                raise wrap_error("This slice is not supported for JAX differentiation.")
+                raise wrap_error(
+                    TypeError("This slice is not supported for JAX differentiation.")
+                )
             # missing values as any integer other than -1 are extremely rare
```
Comment (agoose77) [inline src/awkward/_slicing.py:533]: Unrelated change.

**Review by jpivarski — APPROVED** [review]
